### PR TITLE
rPackages.hdf5r: fix compile error

### PR DESCRIPTION
**Pull request by Kupac**
## Description of changes
Due to https://github.com/HDFGroup/hdf5/issues/1814
, the package doesn't compile, because it needs `h5cc -show` and `h5cc -showconfig` outputs in the config script. Switching back to hdf5_1_10 solved the problem, albeit it's an older version of the hdf5 lib (but seems to work).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by philipp-baumann** [comment]
confirm `nix-build --pure -A rPackages.rhdf5` compiles successfully on this PR for `darwin-aarch64`

```sh
# ...
/nix/store/k687y4hiraslz65qz8pli1wb5wmcabvn-clang-wrapper-16.0.6/bin/c++ -std=gnu++17 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/nix/store/1hfs8ylpsyfrrakhyvskr4dgycd0l16c-R-4.3.2/lib/R/lib -L/nix/store/hi4fialjfb9wd2f5al4vpzlyh8dz38fz-libcxx-16.0.6/lib -o rhdf5.so H5.o H5A.o H5D.o H5E.o H5F.o H5G.o H5I.o H5L.o H5O.o H5P.o H5R.o H5S.o H5S_extras.o H5T.o H5T_extras.o H5Z.o H5constants.o HandleList.o HandleListWrap.o bit64conversion.o external_filters.o h5dump.o h5ls.o h5testLock.o h5writeDataFrame.o printdatatype.o utils.o wrap.o -lhdf5 -lhdf5 -L/nix/store/1hfs8ylpsyfrrakhyvskr4dgycd0l16c-R-4.3.2/lib/R/lib -lR -lintl -Wl,-framework -Wl,CoreFoundation
installing to /nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1/library/00LOCK-rhdf5/00new/rhdf5/libs
** R
** inst
** byte-compile and prepare package for lazy loading
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_TIME failed, using "C" 
3: Setting LC_MESSAGES failed, using "C" 
4: Setting LC_MONETARY failed, using "C" 
** help
*** installing help indices
*** copying figures
** building package indices
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_TIME failed, using "C" 
3: Setting LC_MESSAGES failed, using "C" 
4: Setting LC_MONETARY failed, using "C" 
** installing vignettes
** testing if installed package can be loaded from temporary location
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_TIME failed, using "C" 
3: Setting LC_MESSAGES failed, using "C" 
4: Setting LC_MONETARY failed, using "C" 
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_TIME failed, using "C" 
3: Setting LC_MESSAGES failed, using "C" 
4: Setting LC_MONETARY failed, using "C" 
** testing if installed package keeps a record of temporary installation path
* DONE (rhdf5)
Running phase: fixupPhase
checking for references to /private/tmp/nix-build-r-rhdf5-2.46.1.drv-0/ in /nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1...
patching script interpreter paths in /nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1
rewriting symlink /nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1/nix-support/propagated-user-env-packages to be relative to /nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1
/nix/store/5q2cqvk4ixvpbj9wkni9z1w6l5vk0h9d-r-rhdf5-2.46.1
```

**Comment by b-rodrigues** [comment]
Built and tested successfully on linux-aarch64 (rpi5):

```
> hdf5r::h5version()
hdf5r version 1.3.8 with C-library HDF5 Version  1.10.11 
[1] "1.10.11"
```

**Review by b-rodrigues — APPROVED** [review]

**Comment by b-rodrigues** [comment]
For curiosity's sake, if we wished to be on the latest version, would we need to overlay this definition? https://github.com/NixOS/nixpkgs/blob/cbc4211f0afffe6dfd2478a62615dd5175a13f9a/pkgs/tools/misc/hdf5/1.10.nix#L15

**Comment by Kupac** [comment]
> For curiosity's sake, if we wished to be on the latest version, would we need to overlay this definition?
For curiosity's sake, I'm overriding the version and src of hdf5_1_10 with those of hdf5. if that works, we'd have the latest version compiled without cmake - best of both worlds. And we wouldn't have to update hash at each version bump ;) Let's see how that works !

**Comment by Kupac** [comment]
```
> h5version()
hdf5r version 1.3.8 with C-library HDF5 Version  1.14.3 
[1] "1.14.3"
```

**Comment by Kupac** [comment]
I'm gonna push it, and see what @jbedo  thinks

**Comment by b-rodrigues** [comment]
Everything seems to be working (tested on linux-aarch64):

```
> hdf5r::h5version()
hdf5r version 1.3.8 with C-library HDF5 Version  1.14.3 
[1] "1.14.3"

```

File: pkgs/development/r-modules/default.nix [hunk 1]
```diff
@@ -472,7 +472,10 @@ let
     VariantAnnotation = with pkgs; [ zlib.dev curl.dev ];
     snpStats = [ pkgs.zlib.dev ];
     vcfppR = [ pkgs.curl.dev pkgs.bzip2 pkgs.zlib.dev pkgs.xz];
-    hdf5r = [ pkgs.hdf5.dev ];
+    hdf5r = [ (pkgs.hdf5_1_10.dev.overrideAttrs (prev: new: {
```
Comment (b-rodrigues) [inline pkgs/development/r-modules/default.nix:475]: I think if we keep this, this should likely go all the way down starting at line 944 (where packages with overrides are defined)

**Review by b-rodrigues — CHANGES_REQUESTED** [review]

**Comment by jbedo** [comment]
Fixed in #298837 by leveraging `Rhdf5lib`'s library. If we need a newer version we can work on updating `Rhdf5lib`.

**Pull request closed without merging**